aggregator_core/src/datastore.rs

```diff
@@ -5447,10 +5452,19 @@ mod tests {
         time::Duration as StdDuration,
     };
 
+    #[template]
+    #[rstest]
+    // The version numbers in these cases must match SUPPORTED_SCHEMA_VERSIONS
+    #[case::version_20230405185602(ephemeral_datastore_max_schema_version(20230405185602))]
+    #[case::version_20230417204528(ephemeral_datastore_max_schema_version(20230417204528))]
```

Comment:
Ideally this list of cases would be automatically stamped out from `SUPPORTED_SCHEMA_VERSIONS`, but I am not sure how to do that gracefully.

Reply:
As declarative macros cannot produce attributes, we'd have to write our own procedural macro in order to do that. We can either do so as a progressive enhancement later, or just make coupled changes to the template and `SUPPORTED_SCHEMA_VERSIONS` by hand.

Reply:
I think the main value of `rstest` is that we run each datastore test on each supported schema version. However, I don't like that we have to remember to manually add a case each time we add a new schema version -- we should have one canonical list of supported schemas, and that is `SUPPORTED_SCHEMAS`. Currently, if I add a version to `SUPPORTED_SCHEMAS` but forget to add a case here, the tests would pass even if the new schema versions was broken; I (or a reviewer) have to manually notice that I'm missing a case. IMO, that's a blocker.

A few potential solutions come to mind:

* I assume something like `#[values(SUPPORTED_SCHEMAS)]` doesn't work, but if so, that's easy enough ([docs](https://docs.rs/rstest/0.17.0/rstest/#creating-a-test-for-each-combinations-of-given-values)).
* We could write a macro to generate the necessary `#[case]` attributes -- based on [this StackOverflow question](https://stackoverflow.com/questions/61449167/generate-attribute-value-via-macro) it looks like generating attributes from a macro is possible.
* If the above aren't feasible/easy, another approach would be to avoid using `rstest` at all. We could write a "wrapper function" like `run_with_ephemeral_datastore` which would take a `Fn(EphemeralDatastore)`, then call that function in a loop with an ephemeral datastore at the proper version. Every place that currently uses the template would be changed to call this function instead, wrapping the current test body. This would work, but wouldn't separate different versions of the same test into separate cases or provide parallelization of tests across versions, and less importantly would induce an additional level of indentation in most test bodies.

I would say writing a quick macro is probably the best approach, assuming there's no syntax that would allow `SUPPORTED_VERSIONS` to be used directly.

Reply:
That inspires one way we could do this with a declarative macro: we could define and then use a macro that takes multiple schema versions as arguments, and then emits two top-level items: the `const` declaration of `SUPPORTED_SCHEMA_VERSIONS`, and the entire test template function, decorated with one `#[case]` attribute macro per macro argument.

Reply:
That's a great idea. I'll give the macro a try this afternoon.

Reply:
The macro was pretty easy to write (I did in the latest pushed commit). The downside is that [because you can't construct identifiers with macro variables](https://github.com/rust-lang/rust/issues/29599), we can't use `rstest`'s `case::label` syntax to interpolate the schema version into test names.